catroid/src/main/java/org/catrobat/catroid/transfers/project/ProjectUpload.kt

```diff
@@ -0,0 +1,133 @@
+    private val sceneNames: Array<String>?,
+    private val archiveDirectory: File,
+    private val zipArchiver: ZipArchiver,
+    private val screenshotLoader: ProjectAndSceneScreenshotLoader,
+    private val sharedPreferences: SharedPreferences,
+    private val serverCalls: ServerCalls
+) {
+
+    fun start(
+        successCallback: UploadProjectSuccessCallback,
+        errorCallback: UploadProjectErrorCallback
+    ) {
+        val projectArchive = zipProjectToArchive(projectDirectory, archiveDirectory)
+        if (projectArchive == null) {
+            errorCallback(UPLOAD_ZIP_ERROR, UPLOAD_ZIP_ERROR_MESSAGE)
+            return
+        }
+
+        val projectUploadData = createUploadData(projectArchive)
+
+        scaleSceneScreenshots(projectName, sceneNames)
+
+        serverCalls.uploadProject(
+            projectUploadData,
+            { projectId, successUsername, successToken ->
+                sharedPreferences.edit()
+                    .putString(Constants.TOKEN, successToken)
+                    .putString(Constants.USERNAME, successUsername)
+                    .commit()
+
+                successCallback(projectId)
+                projectArchive.delete()
+            },
+            { errorCode, errorMessage ->
+                errorCallback(
+                    errorCode,
+                    errorMessage
+                )
+            }
+        )
+    }
+
+    private fun createUploadData(projectArchive: File): ProjectUploadData {
+        val token = sharedPreferences.getString(Constants.TOKEN, Constants.NO_TOKEN)
+        val username = sharedPreferences.getString(Constants.USERNAME, Constants.NO_USERNAME)
+
+        return ProjectUploadData(
+            projectName = projectName,
+            projectDescription = projectDescription,
+            projectArchive = projectArchive,
+            userEmail = userEmail,
+            language = Locale.getDefault().language,
+            token = token,
+            username = username
+        )
+    }
+
+    private fun scaleSceneScreenshots(projectName: String, sceneNames: Array<String>?) {
+        sceneNames?.mapNotNull { screenshotLoader.getScreenshotFile(projectName, it, false) }
+            ?.filter { it.exists() && it.length() > 0 }
+            ?.forEach {
+                try {
+                    ImageEditing.scaleImageFile(it, UPLOAD_IMAGE_SCALE_WIDTH, UPLOAD_IMAGE_SCALE_HEIGHT)
+                } catch (ex: FileNotFoundException) {
+                    Log.e(TAG, Log.getStackTraceString(ex))
+                }
+            }
+    }
+
+    private fun zipProjectToArchive(projectDirectory: File, archiveDirectory: File): File? {
+        return try {
+            zipArchiver.zip(archiveDirectory, projectDirectory.listFiles())
+            archiveDirectory
+        } catch (ioException: IOException) {
+            Log.e(TAG, Log.getStackTraceString(ioException))
+            archiveDirectory.delete()
+            null
```

Comment:
Since `zipProjectToArchive` is a private function, it is only used locally in this file. In this particular case, I would prefer to not rethrow another exception since `ProjectUpload` is designed to communicate with surrounding classes via the passed success and error callbacks. What benefit compared to the current implementation would a custom exception provide? We currently do pass an error message in the error callback that tells the user exactly what went wrong.